packages/paymasters/src/TokenPaymasterProvider.ts

```diff
@@ -0,0 +1,129 @@
+import BN from 'bn.js'
+import { RelayClient, RelayProvider, GSNUnresolvedConstructorInput } from '@opengsn/provider'
+import { PrefixedHexString, toChecksumAddress } from 'ethereumjs-util'
+import { Address, removeHexPrefix, Web3ProviderBaseInterface } from '@opengsn/common/dist'
+import { GSNConfig } from '@opengsn/provider/dist/GSNConfigurator'
+import {
+  PermitERC20UniswapV3PaymasterInstance,
+  PermitInterfaceDAIInstance,
+  PermitInterfaceEIP2612Instance
+} from '../types/truffle-contracts'
+import { RelayRequest } from '@opengsn/common/dist/EIP712/RelayRequest'
+import {
+  PERMIT_SIGHASH_DAI, PERMIT_SIGHASH_EIP2612,
+  PERMIT_SIGNATURE_DAI,
+  PERMIT_SIGNATURE_EIP2612,
+  signAndEncodeDaiPermit,
+  signAndEncodeEIP2612Permit
+} from './PermitPaymasterUtils'
+import { constants } from '@opengsn/common/dist/Constants'
+import { EIP712Domain, EIP712DomainType, EIP712DomainTypeWithoutVersion } from '@opengsn/common/dist/EIP712/TypedRequestData'
+import { TokenPaymasterInteractor } from './TokenPaymasterInteractor'
+
+export interface TokenPaymasterConfig extends GSNConfig {
+  tokenAddress: Address
+  permitAmount?: string | number | BN
+}
+
+export interface TokenPaymasterUnresolvedConstructorInput extends GSNUnresolvedConstructorInput {
+  config: Partial<TokenPaymasterConfig>
+}
+
+export class TokenPaymasterProvider extends RelayProvider {
+  config!: TokenPaymasterConfig
+  token!: PermitInterfaceDAIInstance | PermitInterfaceEIP2612Instance
+  permitSignature!: string
+  protected paymaster!: PermitERC20UniswapV3PaymasterInstance
+  readonly tokenPaymasterInteractor: TokenPaymasterInteractor
+
+  constructor (relayClient: RelayClient, provider: Web3ProviderBaseInterface) {
+    super(relayClient)
+    this.tokenPaymasterInteractor = new TokenPaymasterInteractor(provider)
+  }
+
+  static newProvider (input: TokenPaymasterUnresolvedConstructorInput): TokenPaymasterProvider {
+    if (input.config.maxPaymasterDataLength == null) {
+      input.config.maxPaymasterDataLength = 300
+    }
+    const provider = new TokenPaymasterProvider(new RelayClient(input), input.provider)
+    return provider
+  }
+
+  async init (): Promise<this> {
+    await super.init(true)
+    this.relayClient.dependencies.asyncPaymasterData = this._buildPaymasterData.bind(this)
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    this.paymaster = await this.tokenPaymasterInteractor._createPermitERC20UniswapV3Paymaster(this.relayClient.config.tokenPaymasterAddress!)
+    if (this.config.tokenAddress != null) {
+      await this.useToken(this.config.tokenAddress)
+    }
```

Comment:
How would you initialize a `TokenPaymasterProvider` without selecting a token? This should not be an option, I think.

Reply:
Use-case: initialize Provider and select token on the go. Niche but viable.